java/trace/src/main/scala/org/typelevel/otel4s/java/trace/TracerImpl.scala

```diff
@@ -17,65 +17,58 @@
 package org.typelevel.otel4s.java.trace
 
 import cats.effect.Sync
-import cats.mtl.Ask
-import cats.syntax.flatMap._
-import cats.syntax.functor._
 import io.opentelemetry.api.trace.{Span => JSpan}
 import io.opentelemetry.api.trace.{Tracer => JTracer}
 import org.typelevel.otel4s.ContextPropagators
 import org.typelevel.otel4s.TextMapGetter
 import org.typelevel.otel4s.TextMapUpdater
-import org.typelevel.otel4s.context.AskVault
+import org.typelevel.otel4s.java.context.Context
+import org.typelevel.otel4s.java.context.LocalContext
 import org.typelevel.otel4s.trace.SpanBuilder
 import org.typelevel.otel4s.trace.SpanContext
 import org.typelevel.otel4s.trace.Tracer
-import org.typelevel.vault.Vault
 
-private[java] class TracerImpl[F[_]: Sync: AskVault](
+private[java] class TracerImpl[F[_]: Sync](
     jTracer: JTracer,
-    scope: TraceScope[F],
-    propagators: ContextPropagators[F]
-) extends Tracer[F] {
+    propagators: ContextPropagators[F, Context]
+)(implicit L: LocalContext[F])
+    extends Tracer[F] {
 
-  private val runner: SpanRunner[F] = SpanRunner.span(scope)
+  private val runner: SpanRunner[F] = SpanRunner.fromLocal
 
   val meta: Tracer.Meta[F] =
     Tracer.Meta.enabled
 
   def currentSpanContext: F[Option[SpanContext]] =
-    scope.current.map {
-      case Scope.Span(_, jSpan) if jSpan.getSpanContext.isValid =>
-        Some(new WrappedSpanContext(jSpan.getSpanContext))
-
-      case _ =>
-        None
+    L.reader {
+      case Context.Noop => None
+      case Context.Wrapped(underlying) =>
+        Option(JSpan.fromContextOrNull(underlying))
+          .map(jSpan => new WrappedSpanContext(jSpan.getSpanContext))
     }
 
   def spanBuilder(name: String): SpanBuilder[F] =
-    new SpanBuilderImpl[F](jTracer, name, scope, runner)
+    new SpanBuilderImpl[F](jTracer, name, runner)
 
   def childScope[A](parent: SpanContext)(fa: F[A]): F[A] =
-    scope
-      .makeScope(JSpan.wrap(WrappedSpanContext.unwrap(parent)))
-      .flatMap(_(fa))
+    L.local(fa) {
+      _.map(JSpan.wrap(WrappedSpanContext.unwrap(parent)).storeInContext)
+    }
 
   def rootScope[A](fa: F[A]): F[A] =
-    scope.rootScope.flatMap(_(fa))
+    L.local(fa) {
+      case Context.Noop       => Context.Noop
+      case Context.Wrapped(_) => Context.root
+    }
```

Comment:
> I have a feeling that we are trying to squeeze two concepts into one entity: scope management (former `TraceScope`) and context storage (e.g. `Vault`).

@iRevive I don't think the changes to `Context`/`Vault` really touch `TraceScope` directly—this PR basically just replaces `TraceScope` with direct calls to `Local`. with no longer having to worry about storing in a `JContext` and then storing that in a `Vault`, the `TraceScope` implementation simplifies a lot, and it didn't seem worth having a trait for anymore. the only reason I can personally see for keeping `TraceScope` is to keep all the operations in one file so handling of any sharp edges (like having to match for `Noop` in `rootScope`) can be kept together with comments, possibly making mistakes less likely. but I'm not strongly convinced by that argument